Update CLI to not reformat input files

ericcornelissen opened the pull request:
Closes #153

Adjust the implementation of the CLI to leverage the `actions` output list from the main library instead of the `workflow` output to rewrite the input file. In particular, this constructs are regular expression to find `uses:` directives to replace and replaces them (with a fixed format). The result is that the input file isn't reformatted, apart from the `uses:` directives.

Currently this is implemented in for the CLI only. It might make sense to lift this into the main library instead and use this rewrite strategy for the `workflow` output instead of using [`workflow.toString()`](https://github.com/mheap/pin-github-action/blob/8e271c1eb28e643569f502d4df589ce5d77add4e/index.js#L52).

Also note that this duplicates the logic for the default comment from [replaceActions.js](https://github.com/mheap/pin-github-action/blob/8e271c1eb28e643569f502d4df589ce5d77add4e/replaceActions.js#L93-L96). This should probably be addressed before this is merged, but I'm unsure how to resolve that within the current project structure.

lucasgonze commented:
Let's brainstorm on how a regex could backfire, given the prevailing wisdom on [not parsing subtle stuff as strings](https://softwareengineering.stackexchange.com/questions/113237/when-you-should-not-use-regular-expressions).

ericcornelissen commented:
Fair enough @lucasgonze. First of all, if you have a simpler alternative I'm more than happy to use it. Some arguments/thoughs in favor of using regular expressions:

- We know the input is valid YAML because we've already parsed it as YAML. I think it's also safe to assume the input is a file that GitHub Actions can work with (if it isn't I wouldn't expect this program to work properly).
- The motivation for not just serializing the updated data to YAML seems clear to me from #153, though it might also be reasonable to support both approaches and let the user choose.
- Simple substring-based replacement won't do because there's no restrictions on whitespace in YAML files (and I don't think it's reasonable to require users to use a specific amount of whitespace in their workflows).
- This approach (and others, like substring-based replacement) may replace something that shouldn't have been replaced (trivial example: `key: uses: foo/bar@baz`). I would argue that we can generally expect workflows/manifests not to contain such strings. To minimize this risk I intentionally included the `uses:(\\s*)` prefix.
- I believe the syntax for [`uses:`](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstepsuses) is simple/regular enough to be parsed with a regular expressions. Though admittedly, GitHub Actions might be more permissive than I'm thinking.
- I would argue the regular expression isn't very complex, it's mostly used for whitespace matching. That being said, it may be possible to simplify the expression (though what one considers simpler depends on which aspects of the RegExp syntax you're familiar with)
  - A little more about the RegExp: The first `(\\s*)` is forced because there is no restriction on the amount of whitespace after the `:`. The group after `${actionId}@${currentVersion}` is forced because we may need to add a comment to an action that doesn't have it. This group needs to match a comment, which might be preceded by whitespace (so the second `(\\s*)` is forced too). The remaining part is for matching the line comment text, so I opted for the most general expression (anything that's not a line-ending) to match and replace it (which is the same as the existing behavior).
- ReDoS might be a concern for some users (though in most cases I would argue the input is trusted). But this expression is simple enough to avoid ReDoS vulnerabilities, and [ReDoS detector agrees](https://redosdetector.com/?pattern=uses%3A%28%5Cs*%29%24%7BactionId%7D%40%24%7BcurrentVersion%7D%28%5Cs*%23%5B%5E%5Cn%5D*%29%3F&caseInsensitive=false&unicode=false).
- As long as there's concerns I'm more then happy to put this behind an opt-in experimental flag like `--experimental-replace-inline`.

Some example cases to consider (excluding combinations of scenarios):

```yaml
# with leading hyphen
- uses: foo/bar@baz

# without leading hyphen
- name: name
  uses: foo/bar@baz

# no space after uses
- uses:foo/bar@baz

# one space after uses
- uses: foo/bar@baz

# with pin comment, no spacing
- uses: foo/bar@baz#pin@main

# with pin comment, spacing before #
- uses: foo/bar@baz #pin@main

# with pin comment, spacing after #
- uses: foo/bar@baz# pin@main

# with pin comment, spacing before and after #
- uses: foo/bar@baz # pin@main

# with non-pin comment
- uses: foo/bar@baz # foobar
```

Some cases the current expression doesn't handle (should probably be fixed, only realized while writing this response):

- whitespace before colon

  ```yaml
  - uses :foo/bar@baz
  - uses : foo/bar@baz
  ```

- no final newline:

  ```yaml
  - uses: foo/bar@baz
  # No newline at end of file
  ```

ericcornelissen commented:
> Some cases the current expression doesn't handle (should probably be fixed, only realized while writing this response):

These, and one more case, are now handled with e28bbefa1651bf688ec9c2e8e3f1540224bf70e2

mheap commented:
@ericcornelissen I think this is a good addition. For it to be accepted, it would need to be refactored in to `replaceActions` (https://github.com/mheap/pin-github-action/blob/main/replaceActions.js) and tests added

mheap commented:
I’d be ok replacing the implementation so long as all of the tests still pass.On 17 Sep 2024, at 18:20, Eric Cornelissen ***@***.***> wrote:﻿
@mheap, thanks! Before I do that, do you want me to replace the existing implementation or add it as an alternative strategy? And in case of the latter, any preferences for the API?

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

ericcornelissen commented:
@mheap I updated the PR now. Looking forward to your review :slightly_smiling_face: 

I replaced the existing implementation for the new one. The test suite for `replaceActions` still passed (after adjusting for the changed API). I expanded it with the cases I have mentioned in this thread and two more related to indentation.

Also, this change eliminated the need for the `--yaml-line-width` and `--yaml-null-str` inputs I kept them on the CLI to avoid breaking users that expect those flags to exist - I just changed the help text to mark them as deprecated.

ericcornelissen commented on replaceActions.js:
```diff
@@ -1,93 +1,31 @@
-    .items;
-  const steps = runs.filter((n) => n.key == "steps")[0].value.items;
+  const regexpUpdate = new RegExp(
+    `uses(\\s*):(\\s+)${actionId}@${currentVersion}(\\s*)#[^\\n]*`,
```
I was working on something related recently and started using [escape-string-regexp](https://www.npmjs.com/package/escape-string-regexp) (to sanitize the dynamic parts of the expression, `actionId` and `currentVersion` in this case), it might make sense to use it here too - what do you think?

lucasgonze replied:
Worth a try

ericcornelissen replied:
Added in 2947769098f24a7f245d7cdbcf7bacdb257fa2ba. I had to use v4 of the library because v5 (latest) is only available as ESM which is not compatible with this project (relatedly, see https://github.com/mheap/pin-github-action/pull/165#issuecomment-2356087966).

mheap commented:
Thank you for this! It slipped through the cracks somehow, and I just found it whilst doing some housekeeping.

I'll release as a new major version as I've also removed the `--yaml` options

MPV commented:
After upgrading to v2 with this feature (that I've been looking forward to 🙏🙇), I unfortunately started getting failures, with:

`The "data" argument must be of type string or an instance of Buffer, TypedArray, or DataView. Received an instance of Document`
